### PR TITLE
Allow hiding the generic device VID/PID

### PR DESCRIPTION
**Pull request by hughsie**
This hides the "generic" VID/PID that should never be used to match firmware, and can ensure that the customer ID or other vendor-specific identifier has to be used instead.

This ensures that a vendor is unable to push a "generic" firmware that matches against other customer-specific firmware.

Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [X] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

File: plugins/genesys/README.md [hunk 1]
```diff
@@ -28,27 +28,22 @@ This plugin supports the following protocol IDs:
...
+* `USB\VID_03F0&PID_0610&IC_352330&BONDING_0F`
+* `USB\VID_03F0&PID_0610&VENDOR_GENESYSLOGIC&IC_352330&BONDING_0F&PORTNUM_23&VENDORSUP_C09B5DD3-1A23-51D2-995A-F7366AAB3CA4`
+* `USB\VID_05E3&PID_0630&PROJECT_1885D34D-0418-5EF8-8E69-4CEF77B6B6E8`
+* `USB\VID_03F0&PID_0610&PUBKEY_AB859399-95B8-5817-B521-9AD8CC7F5BD6`
 
 These devices also use custom GUID values for the Scaler, e.g.
 
-* HP M24fd USB-C Monitor: `GENESYS_SCALER\MSTAR_TSUM_G&PUBKEY_B335BDCE-7073-5D0E-9BD3-9B69C1A6899F&PANELREV_RIM101`
-* HP M27fd USB-C Monitor: `GENESYS_SCALER\MSTAR_TSUM_G&PUBKEY_847A3650-8648-586B-83C8-8B53714F37E3&PANELREV_RIM101`
+* `GENESYS_SCALER\MSTAR_TSUM_G&PUBKEY_B335BDCE-7073-5D0E-9BD3-9B69C1A6899F&PANELREV_RIM101`
```
Comment (d4s) [inline plugins/genesys/README.md:46]: Why only one monitor left here?

Comment (hughsie) [reply plugins/genesys/README.md:46]: It's supposed to be an example of structure -- not an enumerated list.

File: libfwupdplugin/fu-quirks.h [hunk 2]
```diff
@@ -430,3 +430,17 @@ fu_quirks_add_possible_key(FuQuirks *self, const gchar *possible_key);
  * Since: 1.8.2
  **/
 #define FU_QUIRKS_CFI_DEVICE_BLOCK_SIZE "CfiDeviceBlockSize"
+/**
+ * FU_QUIRKS_SHOW_GUID
```
Comment (d4s) [inline libfwupdplugin/fu-quirks.h:434]: Doesn't that need to mention somewhere "true by default" and/or should be explicitly set to "false"?

Comment (hughsie) [reply libfwupdplugin/fu-quirks.h:434]: I could flip it around -- i.e. make the key `HideGuid=true` -- @superm1 any better names? `InstanceIdOnlyQuirk=true` perhaps?

Comment (d4s) [reply libfwupdplugin/fu-quirks.h:434]: mmm... how about `HideGenericGuid` literally?

Comment (superm1) [reply libfwupdplugin/fu-quirks.h:434]: How about `ShowOnlyPluginGuids`?

**Comment by superm1** [comment]
So another thing that could be done is make it a PLUGIN flag that all devices require requirements. Then any guid can really be matched .

**Comment by hughsie** [comment]
> make it a PLUGIN flag that all devices require requirements

I don't think it can be global; what I want to happen is that silicon vendors switch to using the OEM VID/PID -- which means it's perfectly fine to depend on the generic VID/PID.

File: libfwupdplugin/fu-quirks.h (hunk 2 above)
Comment (hughsie) [reply libfwupdplugin/fu-quirks.h:434]: > ShowOnlyPluginGuids

Nah, we're only hiding the specific one that we match in the quirk file. @d4s is right that it's the wrong way around, so I just flipped it around to `HideGuid`.

**Review by MarekPieta — DISMISSED** [review]

File: plugins/nordic-hid/nordic-hid.quirk [hunk 3]
```diff
@@ -4,3 +4,6 @@ Plugin = nordic_hid
 GType = FuNordicHidCfgChannel
 NordicHidBootloader = B0
 NordicHidGeneration = default
+
+[HIDRAW\VEN_1915&DEV_52DE]
```
[the remark this answers is not in the document]
Comment (hughsie) [reply plugins/nordic-hid/nordic-hid.quirk:8]: Ya, you're right -- I need to rework this and only allow the new GUIDs and none of the generic ones. Will ponder and respin.

[the remark this answers is not in the document]
Comment (hughsie) [reply plugins/nordic-hid/nordic-hid.quirk]: What do you think about the new version?

**Review by adamgene — DISMISSED** [review]
tested and list below.

<details>
<summary>no hide</summary>

```
01:47:19.688 FuMain               FuGenesysUsbhubDevice:
  DeviceId:             b0a78eb71f4eeea7df8fb114522556ba8ce22074
  Name:                 USB2.1 Hub
  Guid:                 0b47e767-d727-59a3-b7ab-92a3515d8f71 ← USB\VID_05E3&PID_0630
  Guid:                 49318c48-6d84-5963-92e5-453212356f7c ← USB\CLASS_09
  Guid:                 c059f9a6-f12a-5960-8e74-ca66ea22b934 ← USB\CLASS_09&SUBCLASS_00
  Guid:                 317baf53-b789-59f5-b08e-aebcb14fac6f ← USB\CLASS_09&SUBCLASS_00&PROT_01
  Guid:                 7bac9b06-d7af-521f-8331-cb2c8995add0 ← USB\CLASS_09&SUBCLASS_00&PROT_02
  Guid:                 57fc80bd-e53a-573d-8d76-728a2f787adf ← USB\VID_05E3&PID_0630&IC_352510
  Guid:                 78d2d4f5-98de-5a3e-894d-d0487ae51100 ← USB\VID_05E3&PID_0630&IC_352510&BONDING_F8
  Guid:                 495060b4-09f4-5c0c-aa9d-9f9f7ce55b90 ← USB\VID_05E3&PID_0630&VENDOR_GENESYSLOGIC&IC_352510&BONDING_F8&PORTNUM_48&VENDORSUP_55283F36-3A46-535D-BBEF-F4706A5536E6
  Guid:                 96b4c772-f531-55dc-87dc-11ac566bc488 ← USB\VID_05E3&PID_0630&PROJECT_BE0F5C39-41D5-55D9-88F0-681418FDEF19
  Plugin:               genesys
  Protocol:             com.genesys.usbhub
  Flags:                updatable|registered|can-verify|can-verify-image|unsigned-payload
  Vendor:               GenesysLogic
  VendorId:             USB:0x05E3
  Version:              89.4
  VersionFormat:        bcd
  VersionRaw:           0x00008904
  InstallDuration:      9
  Created:              2023-11-07

```
</details>
<details>
<summary>hide</summary>

```
01:55:08.430 FuMain               FuGenesysUsbhubDevice:
  DeviceId:             b0a78eb71f4eeea7df8fb114522556ba8ce22074
  Name:                 USB2.1 Hub
  Guid:                 57fc80bd-e53a-573d-8d76-728a2f787adf ← USB\VID_05E3&PID_0630&IC_352510
  Guid:                 78d2d4f5-98de-5a3e-894d-d0487ae51100 ← USB\VID_05E3&PID_0630&IC_352510&BONDING_F8
  Guid:                 495060b4-09f4-5c0c-aa9d-9f9f7ce55b90 ← USB\VID_05E3&PID_0630&VENDOR_GENESYSLOGIC&IC_352510&BONDING_F8&PORTNUM_48&VENDORSUP_55283F36-3A46-535D-BBEF-F4706A5536E6
  Guid:                 96b4c772-f531-55dc-87dc-11ac566bc488 ← USB\VID_05E3&PID_0630&PROJECT_BE0F5C39-41D5-55D9-88F0-681418FDEF19
  Plugin:               genesys
  Protocol:             com.genesys.usbhub
  Flags:                updatable|registered|can-verify|can-verify-image|unsigned-payload
  Vendor:               GenesysLogic
  VendorId:             USB:0x05E3
  Version:              89.4
  VersionFormat:        bcd
  VersionRaw:           0x00008904
  InstallDuration:      9
  Created:              2023-11-07
```
</details>

**Review by d4s — APPROVED** [review]

**Review by d4s — APPROVED** [review]

**Comment by hughsie** [comment]
@superm1 if this looks okay to you I'll merge and then rebase the other one on top (they'll be conflicts)

**Review by d4s — DISMISSED** [review]
LGTM. Adding flag `FU_DEVICE_INTERNAL_FLAG_NO_GENERIC_GUIDS` works for now.

**Review by MarekPieta — APPROVED** [review]
Confirmed that change works fine for nordic-hid plugin. Code looks good too. Thanks for the changes.

Edit: Seems that you were faster, @hughsie  ;)

**Comment by superm1** [comment]
I tried to add this to amd-gpu and it seems not to work. Does it work for other udev based plugins?

**Comment by hughsie** [comment]
> I tried to add this to amd-gpu and it seems not to work

You need the latest PR -- can you try what I merged to main? If not, can you share a -vv log please.

**Pull request merged**